substrate/frame/delegated-staking/src/lib.rs

```diff
@@ -0,0 +1,763 @@
+
+use types::*;
+
+mod impls;
+
+use frame_support::{
+	pallet_prelude::*,
+	traits::{
+		fungible::{
+			hold::{
+				Balanced as FunHoldBalanced, Inspect as FunHoldInspect, Mutate as FunHoldMutate,
+			},
+			Balanced, Inspect as FunInspect, Mutate as FunMutate,
+		},
+		tokens::{fungible::Credit, Precision, Preservation},
+		Defensive, DefensiveOption, Imbalance, OnUnbalanced,
+	},
+};
+
+use sp_runtime::{
+	traits::{AccountIdConversion, CheckedAdd, CheckedSub, Zero},
+	ArithmeticError, DispatchResult, RuntimeDebug, Saturating,
+};
+use sp_staking::{EraIndex, StakingInterface, StakingUnsafe};
+use sp_std::{convert::TryInto, prelude::*};
+
+pub type BalanceOf<T> =
+	<<T as Config>::Currency as FunInspect<<T as frame_system::Config>::AccountId>>::Balance;
+
+use frame_system::{ensure_signed, pallet_prelude::*, RawOrigin};
+
+#[frame_support::pallet]
+pub mod pallet {
+	use super::*;
+
+	#[pallet::pallet]
+	pub struct Pallet<T>(PhantomData<T>);
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		/// The overarching event type.
+		type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+
+		/// Injected identifier for the pallet.
+		#[pallet::constant]
+		type PalletId: Get<frame_support::PalletId>;
+
+		/// Currency type.
+		type Currency: FunHoldMutate<Self::AccountId, Reason = Self::RuntimeHoldReason>
+			+ FunMutate<Self::AccountId>
+			+ FunHoldBalanced<Self::AccountId>;
+
+		/// Handler for the unbalanced reduction when slashing a delegator.
+		type OnSlash: OnUnbalanced<Credit<Self::AccountId, Self::Currency>>;
+
+		/// Overarching hold reason.
+		type RuntimeHoldReason: From<HoldReason>;
+
+		/// Core staking implementation.
+		type CoreStaking: StakingUnsafe<Balance = BalanceOf<Self>, AccountId = Self::AccountId>;
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		/// The account cannot perform this operation.
+		NotAllowed,
+		/// An existing staker cannot perform this action.
+		AlreadyStaking,
+		/// Reward Destination cannot be same as `Agent` account.
+		InvalidRewardDestination,
+		/// Delegation conditions are not met.
+		///
+		/// Possible issues are
+		/// 1) Cannot delegate to self,
+		/// 2) Cannot delegate to multiple delegates,
+		InvalidDelegation,
+		/// The account does not have enough funds to perform the operation.
+		NotEnoughFunds,
+		/// Not an existing `Agent` account.
+		NotAgent,
+		/// Not a Delegator account.
+		NotDelegator,
+		/// Some corruption in internal state.
+		BadState,
+		/// Unapplied pending slash restricts operation on `Agent`.
+		UnappliedSlash,
+		/// `Agent` has no pending slash to be applied.
+		NothingToSlash,
+		/// Failed to withdraw amount from Core Staking.
+		WithdrawFailed,
+		/// Operation not supported by this pallet.
+		NotSupported,
+	}
+
+	/// A reason for placing a hold on funds.
+	#[pallet::composite_enum]
+	pub enum HoldReason {
+		/// Funds held for stake delegation to another account.
+		#[codec(index = 0)]
+		Delegating,
```

Comment:
Might be confused in a UI with governance delegation, so let's call it `StakingDelegation` or similar?